dashboard/api.py

```diff
@@ -0,0 +1,227 @@
+                cls.UPGRADE, cls.NOT_OFFERED, cls.OFFERED]
+
+
+class CourseRunStatus(object):
+    """
+    Possible statuses for a course run for a user
+    """
+    NOT_ENROLLED = 'not-enrolled'
+    GRADE = 'grade'
+    READ_CERT = 'read-cert'
+    WILL_ATTEND = 'will-attend'
+    UPGRADE = 'upgrade'
+    NOT_PASSED = 'not-passed'
+
+
+class CourseFormatConditionalFields(object):
+    """
+    The formatting of a course run is dependent
+    on the CourseStatus status passed on the function.
+    There are some fields that are common and others
+    that depend on the status. Also the name of the fields changes.
+
+    This class contains the association between the CourseStatus status
+    that need specific fields, the field associated correspondent
+    to a course run and the new name they need to have.
+    """
+    ASSOCIATED_FIELDS = {
+        CourseStatus.OFFERED: [
+            {
+                'course_run_field': 'enrollment_start',
+                'format_field': 'enrollment_start_date'
+            },
+            {
+                'course_run_field': 'fuzzy_enrollment_start_date',
+                'format_field': 'fuzzy_enrollment_start_date'
+            },
+        ],
+        CourseStatus.CURRENT_GRADE: [
+            {
+                'course_run_field': 'start_date',
+                'format_field': 'course_start_date'
+            },
+        ]
+    }
+
+    @classmethod
+    def get_assoc_field(cls, course_status):
+        """
+        Method to get from the ASSOCIATED_FIELDS dict
+        """
+        return cls.ASSOCIATED_FIELDS.get(course_status, [])
+
+
+class CourseRunUserStatus(object):
+    """
+    Representation of a course run status for a specific user
+    """
+    def __init__(self, status, course_run=None, enrollment_for_course=None):
+        self.status = status
+        self.course_run = course_run
+        self.enrollment_for_course = enrollment_for_course
+
+
+def get_info_for_course(user, course, user_enrollments, user_certificates):
+    """
+    Checks the status of a course given the status of all its runs
+
+    Args:
+        user (User): an user object
+        course (Course): a course object
+        user_enrollments (Enrollments): the user enrollments object
+        user_certificates (Certificates): the user certificates objects
+
+    Returns:
+        dict: dictionary representing the course status for the user
+    """
+    # pylint: disable=unused-argument, too-many-return-statements
+    if not course.courserun_set.count():
+        return format_course_for_dashboard(None, CourseStatus.NOT_OFFERED, course)
+    # get all the run statuses
+    run_statuses = []
+    for course_run in course.courserun_set.all():
+        run_statuses.append(get_status_for_courserun(course_run, user_enrollments))
+    # sort them by end date
+    run_statuses.sort(key=lambda x: x.course_run.end_date or datetime.datetime(datetime.MAXYEAR, 1, 1), reverse=True)
```

Comment:
According to the postgres docs it already sorts null values as being greater than any other value (from http://www.postgresql.org/docs/8.3/static/queries-order.html), so you may be able to just sort this with an `order_by` instead